### PR TITLE
feat: cloudflare worker to inject meta tags

### PR DESCRIPTION
**Pull request by LunrEclipse**
## Description

Created Cloudflare Workers to dynamically inject metadata into NFTs, NFT Collections, and Tokens. Created routed functions `/asset/[[index]].ts`, `/tokens/[[index]]/ts`, and `collection/[index]/ts` that utilize GraphQL to query the appropriate information and inject them into the `<head>` using Cloudflare's `HTMLRewriter. 

Unit testing can be found in `collection.test.ts`, `nft.test.ts`, and `token.test.ts`. 

### QA (ie manual testing)

- [x] Manually start Cloudflare test environment using `yarn start:cloud` and navigate to the corresponding routes to check if metadata properly injected. 

### Automated testing

- [x] Unit test `yarn test:cloud`

### CODE REVIEW

File: functions/components/assetInjector.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  image: any
+  uniswapUrl: any
+}
+
+export class MetaTagInjector {
+  private input: MetaTagInjectorInput
+
+  constructor(input: MetaTagInjectorInput) {
+    this.input = input
+  }
```
Comment (zzmp) [inline functions/components/assetInjector.ts:12]:
```suggestion
  constructor(private input: MetaTagInjectorInput) {}
```

File: functions/components/assetInjector.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+  element(element) {
+    //Open Graph Tags
+    element.append(`<meta property="og:title" content = "${this.input.name}"/>`, {
```
Comment (zzmp) [inline functions/components/assetInjector.ts:16]: Here and below, remove the spacing around the content assignment: `content="`

File: functions/components/assetInjector.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
+type MetaTagInjectorInput = {
+  name: any
+  image: any
+  uniswapUrl: any
```
Comment (zzmp) [inline functions/components/assetInjector.ts:4]: This can just be `url` to make it less confusing.

File: functions/components/collectionInjector.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
+type MetaTagInjectorInput = {
+  name: any
+  image: any
+  uniswapUrl: any
+  isVerified: any
```
Comment (zzmp) [inline functions/components/collectionInjector.ts:5]: Is this used?

File: functions/components/tokenInjector.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
+type MetaTagInjectorInput = {
+  name: any
+  network: any
```
Comment (zzmp) [inline functions/components/tokenInjector.ts:3]: Is this used?

File: functions/nfts/asset/[[index]].ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+import getAsset from '../../utils/getAsset'
+
+export const onRequest: PagesFunction = async ({ params, request, next }) => {
+  try {
+    const { index } = params
+    const collectionAddress = String(index[0])
+    const tokenId = String(index[1])
+    const data = await getAsset(collectionAddress, tokenId, request.url)
+    if (!data) {
+      return await next()
```
Comment (zzmp) [inline functions/nfts/asset/[[index]].ts:12]:
```suggestion
      return next()
```

File: functions/nfts/asset/[[index]].ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    }
+    return new HTMLRewriter().on('head', new MetaTagInjector(data)).transform(await next())
```
Comment (zzmp) [inline functions/nfts/asset/[[index]].ts:14]: You can parallelize this, but you'll need to save a reference to `res` to return in case of rejections

```
const assetPromise = getAsset(collectionAddress, tokenId, request.url)
const resPromise = next()
try {
  const [asset, res] = await Promise.all([assetPromise, resPromise])
  ...
} catch (e) {
  return resPromise
}
```

File: functions/utils/getApolloClient.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  return new ApolloClient({
+    connectToDevTools: true,
+    uri: GRAPHQL_ENDPOINT,
+    headers: {
+      'Content-Type': 'application/json',
+      Origin: 'https://app.uniswap.org',
+      'User-Agent':
+        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.5735.110 Safari/537.36',
+    },
+    cache: new InMemoryCache(),
```
Comment (zzmp) [inline functions/utils/getApolloClient.ts:14]: Do you know how caching will work within the worker? I assume it'd be hard to cache across requests.
Will endpoints with the same parameters just be cached for us automatically?

**Review by zzmp — CHANGES_REQUESTED** [review]
It looks like every injector is injecting the same tags with different values. Instead of having three different injectors, which means more code to test, could you write one injector that takes the tag values as input, and then have the handlers convert asset metadata to tag value inputs? This could hopefully reduce your code by a lot.

File: functions/collection.test.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+      `<meta property="og:image:alt" content = "https://app.uniswap.org/images/512x512_App_Icon.png"/>`
+    )
+    expect(body).toContain(`<meta property="twitter:card" content = "summary_large_image"/>`)
+    expect(body).toContain(`<meta property="twitter:title" content = "${nft.collectionName} on Uniswap"/>`)
+    expect(body).toContain(`<meta property="twitter:image" content = "${nft.image}"/>`)
+    expect(body).toContain(
+      `<meta property="twitter:image:alt" content = "https://app.uniswap.org/images/512x512_App_Icon.png"/>`
+    )
+  }
+}, 50000)
```
[the remark this answers is not in the document]
Comment (LunrEclipse) [reply functions/collection.test.ts:48]: Yeah, so on the initial run, yarn takes extra time to compile and setup the proxy, so the 50 second timeout is needed otherwise the tests will just fail as they wait for the server to get started

File: functions/collection.test.ts [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+  for (const nft of collections) {
+    const url = 'http://127.0.0.1:3000/nfts/collection/' + nft.address
+    const req = new Request(url)
+    const res = await fetch(req)
+    const body = await res.text()
+    expect(body).toMatchSnapshot()
+    expect(body).toContain(nft.collectionName + ' on Uniswap')
+    expect(body).toContain(nft.image)
+    expect(body).toContain(url)
+    expect(body).toContain(`<meta property="og:title" content = "${nft.collectionName} on Uniswap"/>`)
```
[the remark this answers is not in the document]
Comment (LunrEclipse) [reply functions/collection.test.ts:32]: Good question! I will remove

File: functions/collection.test.ts [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    expect(body).toContain(`<meta property="twitter:title" content = "${nft.collectionName} on Uniswap"/>`)
+    expect(body).toContain(`<meta property="twitter:image" content = "${nft.image}"/>`)
+    expect(body).toContain(
+      `<meta property="twitter:image:alt" content = "https://app.uniswap.org/images/512x512_App_Icon.png"/>`
+    )
+  }
+}, 50000)
+
+test('should not inject metadata for invalid calls', async () => {
+  const baseReq = new Request('http://127.0.0.1:3000/nfts/collection/0xed5af388653567af2f388e6224dc7c4b3241c545')
```
[the remark this answers is not in the document]
Comment (LunrEclipse) [reply functions/collection.test.ts:51]: The collection address is invalid (i.e. I don't think the contract exists)

File: functions/collection.test.ts [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+  const baseRes = await fetch(baseReq)
+  const baseBody = await baseRes.text()
+  expect(baseBody).toMatchSnapshot()
```
[the remark this answers is not in the document]
Comment (LunrEclipse) [reply functions/collection.test.ts:54]: Not really

File: functions/nfts/asset/[[index]].ts [hunk 13]
```diff
@@ -0,0 +1,20 @@
+/* eslint-disable import/no-unused-modules */
+import { MetaTagInjector } from '../../components/metaTagInjector'
+import getAsset from '../../utils/getAsset'
+
+export const onRequest: PagesFunction = async ({ params, request, next }) => {
+  const { index } = params
+  const collectionAddress = String(index[0])
+  const tokenId = String(index[1])
```
Comment (zzmp) [inline functions/nfts/asset/[[index]].ts:8]: Why do you use `String`? This is pretty non-standard - can you instead do something like `index[0]?.toString()`? Or is it guaranteed to be a string?

Is another alternative to change the filepath of this file to capture the collectionAddress and tokenId?

File: functions/nfts/asset/[[index]].ts [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+  const assetPromise = getAsset(collectionAddress, tokenId, request.url)
+  const resPromise = next()
+  try {
+    const [data, res] = await Promise.all([assetPromise, resPromise])
```
Comment (zzmp) [inline functions/nfts/asset/[[index]].ts:12]: Call it `asset` to match the `assetPromise`.
```suggestion
    const [asset, res] = await Promise.all([assetPromise, resPromise])
```

File: functions/nfts/asset/[[index]].ts [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+    if (!data) {
+      return resPromise
+    }
+    return new HTMLRewriter().on('head', new MetaTagInjector(data)).transform(res)
+  } catch (e) {
+    return resPromise
```
Comment (zzmp) [inline functions/nfts/asset/[[index]].ts:18]: Do we want to `console.error(e)`? Do you know where it will be logged?

File: functions/token.test.ts [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+test.each(tokens)(
+  'should inject metadata for valid tokens',
+  async (token) => {
+    const url = 'http://127.0.0.1:3000/tokens/' + token.network + '/' + token.address
+    const body = await fetch(new Request(url)).then((res) => res.text())
+    expect(body).toMatchSnapshot()
+    expect(body).toContain('Get ' + token.symbol + ' on Uniswap')
+    expect(body).toContain(token.image)
+    expect(body).toContain(url)
```
Comment (zzmp) [inline functions/token.test.ts:46]:
```suggestion
```

File: functions/tokens/[[index]].ts [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+import getToken from '../utils/getToken'
+
+export const onRequest: PagesFunction = async ({ params, request, next }) => {
+  const { index } = params
+  const networkName = String(index[0]).toUpperCase()
+  let tokenAddress = String(index[1])
+  tokenAddress =
+    tokenAddress !== 'undefined' && tokenAddress === 'NATIVE'
+      ? '0x0000000000000000000000000000000000000000'
+      : tokenAddress
```
Comment (zzmp) [inline functions/tokens/[[index]].ts:12]: Instead of using a var (vs const), pull this logic into a `function toTokenAddress` at the top of the file. This has the benefits of (1) being const and (2) being more self-documenting in code.

File: functions/utils/getApolloClient.ts [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+export const getApolloClient = () => {
+  return new ApolloClient({
+    connectToDevTools: true,
+    uri: GRAPHQL_ENDPOINT,
+    headers: {
+      'Content-Type': 'application/json',
+      Origin: 'https://app.uniswap.org',
+      'User-Agent':
+        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.5735.110 Safari/537.36',
+    },
```
Comment (zzmp) [inline functions/utils/getApolloClient.ts:13]: Is there a reason for this specific user-agent?

File: functions/utils/getApolloClient.ts [hunk 19]
```diff
@@ -0,0 +1,21 @@
+import { ApolloClient, InMemoryCache } from '@apollo/client'
+const GRAPHQL_ENDPOINT = 'https://api.uniswap.org/v1/graphql'
+
+export const getApolloClient = () => {
+  return new ApolloClient({
+    connectToDevTools: true,
+    uri: GRAPHQL_ENDPOINT,
+    headers: {
+      'Content-Type': 'application/json',
+      Origin: 'https://app.uniswap.org',
```
Comment (zzmp) [inline functions/utils/getApolloClient.ts:10]: Do you _need_ an origin?

File: functions/utils/getApolloClient.ts [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+      'User-Agent':
+        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.5735.110 Safari/537.36',
+    },
+    cache: new InMemoryCache(),
+    defaultOptions: {
+      watchQuery: {
+        fetchPolicy: 'cache-and-network',
```
Comment (zzmp) [inline functions/utils/getApolloClient.ts:17]: Why not `cache-first`? [docs](https://www.apollographql.com/docs/react/data/queries/#supported-fetch-policies)

File: functions/utils/getApolloClient.ts [hunk 21]
```diff
@@ -0,0 +1,21 @@
+import { ApolloClient, InMemoryCache } from '@apollo/client'
+const GRAPHQL_ENDPOINT = 'https://api.uniswap.org/v1/graphql'
+
+export const getApolloClient = () => {
```
Comment (zzmp) [inline functions/utils/getApolloClient.ts:4]: You can do this in a follow-up, but you should leave a TODO for yourself:
This instantiates a new client for every call, so the in-memory cache will never even be used. You need to store this as a global var. You can probably just `export default new ApolloClient...`, instead of having a `getApolloClient` method.

File: functions/utils/getAsset.ts [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+        tokenIds: [tokenId],
+      },
+    },
+  })
+  const asset = data?.nftAssets?.edges[0]?.node
+  if (!asset) {
+    return undefined
+  }
+  const formattedAsset = {
+    title: asset.name ? asset.name : asset.collection?.name + ' #' + asset.tokenId,
```
Comment (zzmp) [inline functions/utils/getAsset.ts:20]: you should check for `asset.collection?.name` before using it, otherwise you'll get a leading space if it's not there.

File: functions/utils/getAsset.ts [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+    image: asset.image?.url,
```
Comment (zzmp) [inline functions/utils/getAsset.ts:21]: These images may not be available. If it is not available, do we still want to inject meta tags for images?

**Review by zzmp — CHANGES_REQUESTED** [review]

File: functions/collection.test.ts [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+    expect(body).toContain(`<meta property="twitter:title" content="${collection.collectionName} on Uniswap"/>`)
+    expect(body).toContain(`<meta property="twitter:image" content="${collection.image}"/>`)
+    expect(body).toContain(
+      `<meta property="twitter:image:alt" content="https://app.uniswap.org/images/512x512_App_Icon.png"/>`
+    )
+  },
+  50000
+)
+
+const invalidCollections = [
```
[the remark this answers is not in the document]
Comment (LunrEclipse) [reply functions/collection.test.ts:48]: Screenshot 2023-07-13 at 4 33 50 PM

I tried that, but I kept getting redeclaration issues

File: functions/utils/getApolloClient.ts (hunk 20 above)
Comment (LunrEclipse) [reply functions/utils/getApolloClient.ts:17]: `cache-first` would make sense, but also if we're using cloudflare's cache wouldn't `network-only` be better?

File: functions/utils/getApolloClient.ts (hunk 19 above)
Comment (LunrEclipse) [reply functions/utils/getApolloClient.ts:10]: yeah otherwise it gets autofiltered by backend

File: functions/utils/getAsset.ts (hunk 23 above)
Comment (LunrEclipse) [reply functions/utils/getAsset.ts:21]: I think it should be fine. Worst case we just have the default Uniswap image as the fallback image url.

File: functions/components/metaTagInjector.ts [hunk 25]
```diff
@@ -7,44 +7,27 @@ type MetaTagInjectorInput = {
 export class MetaTagInjector {
```
Comment (zzmp) [inline functions/components/metaTagInjector.ts:7]: The `MetaTagInjectorInput` properties should not use `any`, it should require `string`s.

File: functions/components/metaTagInjector.ts [hunk 26]
```diff
@@ -0,0 +1,50 @@
+type MetaTagInjectorInput = {
+  title: any
+  image: any
+  url: any
+}
+
+export class MetaTagInjector {
```
[the remark this answers is not in the document]
Comment (zzmp) [reply functions/components/metaTagInjector.ts:7]: nit: The class should also have a doc-comment.

File: functions/utils/getApolloClient.ts (hunk 20 above)
Comment (zzmp) [reply functions/utils/getApolloClient.ts]: this refers to the in-memory cache (as you've set it), so I think you want both.

File: functions/utils/getAsset.ts (hunk 23 above)
Comment (zzmp) [reply functions/utils/getAsset.ts:24]: but you shouldn't render the tag at all in that case, right?

**Review by zzmp — CHANGES_REQUESTED** [review]
there's a few unresolved comments left - can you look at those, and then this is virtually good to go.

File: functions/utils/getApolloClient.ts (hunk 20 above)
Comment (LunrEclipse) [reply functions/utils/getApolloClient.ts]: yeah your right. i guess we can always change it in the cache pr

File: functions/components/metaTagInjector.ts [hunk 27]
```diff
@@ -17,17 +21,19 @@ export class MetaTagInjector {
...
-    this.append(element, 'og:image', this.input.image)
-    this.append(element, 'og:image:width', '1200')
-    this.append(element, 'og:image:height', '630')
+    if (this.input.image) {
+      this.append(element, 'og:image', this.input.image)
+      this.append(element, 'og:image:width', '1200')
+      this.append(element, 'og:image:height', '630')
+      this.append(element, 'og:image:alt', this.input.title)
+      this.append(element, 'twitter:image', this.input.image)
+    }
```
Comment (zzmp) [inline functions/components/metaTagInjector.ts:30]: Put this with the twitter tags - just do the same `if (this.input.image) {` check down there.

File: functions/utils/getApolloClient.ts [hunk 28]
```diff
@@ -0,0 +1,20 @@
+import { ApolloClient, InMemoryCache } from '@apollo/client'
+const GRAPHQL_ENDPOINT = 'https://api.uniswap.org/v1/graphql'
+
+//TODO: Figure out how to make ApolloClient global variable
+export default new ApolloClient({
```
Comment (zzmp) [inline functions/utils/getApolloClient.ts:5]: this filename is no longer correct. Can it be moved to `functions/client.ts`?

File: functions/utils/getApolloClient.ts [hunk 29]
```diff
@@ -0,0 +1,20 @@
...
+  connectToDevTools: true,
+  uri: GRAPHQL_ENDPOINT,
+  headers: {
+    'Content-Type': 'application/json',
+    Origin: 'https://app.uniswap.org',
+    'User-Agent':
```
Comment (zzmp) [inline functions/utils/getApolloClient.ts:11]: Why is the User-Agent set to this value?

File: functions/tokens/[[index]].ts [hunk 30]
```diff
@@ -0,0 +1,27 @@
...
+  const { index } = params
+  const networkName = index[0]?.toString().toUpperCase()
+  let tokenAddress = index[1]?.toString()
+  if (!tokenAddress) {
+    return next()
+  }
+  tokenAddress =
+    tokenAddress !== 'undefined' && tokenAddress === 'NATIVE'
+      ? '0x0000000000000000000000000000000000000000'
+      : tokenAddress
```
Comment (zzmp) [inline functions/tokens/[[index]].ts:15]: Can you move this to a function at top of file to avoid using `let`?

**Review by zzmp — CHANGES_REQUESTED** [review]

File: functions/utils/getApolloClient.ts (hunk 29 above)
Comment (LunrEclipse) [reply functions/utils/getApolloClient.ts:11]: No particular reason LOL. According to Andrew I just needed a user-agent, so I used the one he sent me and have been using it ever since

File: functions/components/metaTagInjector.ts [hunk 31]
```diff
@@ -0,0 +1,41 @@
...
+    this.append(element, 'og:type', 'website')
+    this.append(element, 'og:url', this.input.url)
+
+    //Twitter Tags
+    this.append(element, 'twitter:card', 'summary_large_image')
+    this.append(element, 'twitter:title', this.input.title)
+    if (this.input.image) {
+      this.append(element, 'twitter:image', this.input.image)
+    }
+    this.append(element, 'twitter:image:alt', this.input.title)
```
Comment (zzmp) [inline functions/components/metaTagInjector.ts:39]:
```suggestion
      this.append(element, 'twitter:image:alt', this.input.title)
    }
```

**Review by zzmp — CHANGES_REQUESTED** [review]
last nit

**Review by zzmp — APPROVED** [review]

File: functions/components/metaTagInjector.ts [hunk 32]
```diff
@@ -0,0 +1,41 @@
+type MetaTagInjectorInput = {
+  title: string
+  image?: string
+  url: string
+}
+
+/**
+ * Wrapper class for Cloudflare's HTMLRewriter {@link https://developers.cloudflare.com/workers/runtime-apis/html-rewriter}
```
Comment (just-toby) [inline functions/components/metaTagInjector.ts:8]: i would rephrase this as it's not really wrapping HTMLRewriter. maybe say it's a "listener" class for HTMLRewriter events. or a "subscriber"

File: functions/tokens/[[index]].ts [hunk 33]
```diff
@@ -0,0 +1,27 @@
...
+
+const convertTokenAddress = (tokenAddress: string) => {
+  return tokenAddress && tokenAddress === 'NATIVE' ? '0x0000000000000000000000000000000000000000' : tokenAddress
+}
+
+export const onRequest: PagesFunction = async ({ params, request, next }) => {
+  const { index } = params
+  const networkName = index[0]?.toString().toUpperCase()
+  const tokenAddress = convertTokenAddress(index[1]?.toString())
+  if (!tokenAddress) {
```
Comment (just-toby) [inline functions/tokens/[[index]].ts:13]: what if `networkName` is not defined?

File: functions/utils/getAsset.ts [hunk 34]
```diff
@@ -0,0 +1,28 @@
...
+  const asset = data?.nftAssets?.edges[0]?.node
+  if (!asset) {
+    return undefined
+  }
+  const formattedAsset = {
+    title: asset.name
+      ? asset.name
+      : asset.collection?.name
+      ? asset.collection.name + ' #' + asset.tokenId
+      : 'Asset #' + asset.tokenId,
```
Comment (just-toby) [inline functions/utils/getAsset.ts:23]: this is a little hard to read. maybe use the "nullish coalescing" operator (i.e. `??`) to cut down on visual branching, or just make a helper function

File: functions/client.ts [hunk 35]
```diff
@@ -0,0 +1,20 @@
+import { ApolloClient, InMemoryCache } from '@apollo/client'
+const GRAPHQL_ENDPOINT = 'https://api.uniswap.org/v1/graphql'
+
+//TODO: Figure out how to make ApolloClient global variable
```
Comment (just-toby) [inline functions/client.ts:4]: what does this mean? are you just trying to share a single instance of the client across files, or something else?

File: functions/utils/getAsset.ts (hunk 34 above)
Comment (just-toby) [reply functions/utils/getAsset.ts:23]: also, maybe we should have a fallback title here if none of the information is found

File: functions/utils/getCollection.ts [hunk 36]
```diff
@@ -0,0 +1,21 @@
...
+  const { data } = await client.query({
+    query: CollectionDocument,
+    variables: {
+      addresses: collectionAddress,
+    },
+  })
+  const collection = data?.nftCollections?.edges[0]?.node
+  if (!collection || !collection.name) {
+    return undefined
+  }
```
Comment (just-toby) [inline functions/utils/getCollection.ts:14]: this behaves differently than the asset one right? should we return undefined in asset if we can't find a name / info ?

File: functions/utils/getToken.ts [hunk 37]
```diff
@@ -0,0 +1,22 @@
...
+      chain: networkName,
+      address: tokenAddress,
+    },
+  })
+  const asset = data?.token
+  if (!asset) {
+    return undefined
+  }
+  const formattedAsset = {
+    title: 'Get ' + asset.symbol + ' on Uniswap',
```
Comment (just-toby) [inline functions/utils/getToken.ts:17]: is symbol always defined?

**Review by just-toby — CHANGES_REQUESTED** [review]
RC for some discussion, overall looks good though

File: functions/tokens/[[index]].ts (hunk 33 above)
Comment (LunrEclipse) [reply functions/tokens/[[index]].ts:13]: I don't believe this function ever gets called if `networkName` is undefined.

File: functions/utils/getCollection.ts (hunk 36 above)
Comment (LunrEclipse) [reply functions/utils/getCollection.ts:14]: From what I've looked into, a lot of assets don't actually have names which is why I had to create the weird branching naming structure. For collections we can do this since for some reason the GraphQL indexes every single valid address as an `NFTContract` even if that is not the case, so the only way we can differentiate whether or not a collection is in fact an `NFTContract` is through the name.

File: functions/utils/getToken.ts (hunk 37 above)
Comment (LunrEclipse) [reply functions/utils/getToken.ts:17]: It should be as symbol is part of the constructor for ERC20 contracts, but then again someone can always deployed a modified version so I guess not always.

**Review by just-toby — APPROVED** [review]
